linkerd/app/core/src/proxy/server.rs

```diff
@@ -34,35 +38,43 @@ pub type Connection = (Protocol, BoxedIo);
 
 #[derive(Clone, Debug)]
 pub struct ProtocolDetect {
+    capacity: usize,
     skip_ports: Arc<IndexSet<u16>>,
 }
 
 impl ProtocolDetect {
+    const PEEK_CAPACITY: usize = 8192;
+
     pub fn new(skip_ports: Arc<IndexSet<u16>>) -> Self {
-        ProtocolDetect { skip_ports }
+        ProtocolDetect {
+            skip_ports,
+            capacity: Self::PEEK_CAPACITY,
+        }
     }
 }
 
+#[async_trait]
 impl detect::Detect<tls::accept::Meta> for ProtocolDetect {
     type Target = Protocol;
+    type Error = io::Error;
 
-    fn detect_before_peek(
+    async fn detect(
         &self,
         tls: tls::accept::Meta,
-    ) -> Result<Self::Target, tls::accept::Meta> {
+        io: BoxedIo,
+    ) -> Result<(Self::Target, BoxedIo), Self::Error> {
         let port = tls.addrs.target_addr().port();
         if self.skip_ports.contains(&port) {
-            return Ok(Protocol { tls, http: None });
+            let proto = Protocol { tls, http: None };
+            return Ok::<_, Self::Error>((proto, io));
         }
 
-        Err(tls)
-    }
-
-    fn detect_peeked_prefix(&self, tls: tls::accept::Meta, prefix: &[u8]) -> Self::Target {
-        Protocol {
-            tls,
-            http: HttpVersion::from_prefix(prefix),
-        }
+        // Otherwise, attempt to peek the client connection to determine the protocol.
```

Comment:
nit, take it or leave it: it's a _little_ weird to me that this comment starts with "otherwise" when there's no preceding comment. clearly it's intended to be read as "if self.skip_ports.contains(&port), ..., otherwise, ...", but the "otherwise" made me look for another comment before i figured that out. this probably definitely doesn't actually matter.